synapse/http/matrixfederationclient.py

```diff
@@ -188,6 +188,69 @@ def schedule(x):
 
         self._cooperator = Cooperator(scheduler=schedule)
 
+    @defer.inlineCallbacks
+    def _send_request_with_optional_trailing_slash(
+        self,
+        request,
+        try_trailing_slash_on_400=False,
+        backoff_on_404=False,
+        send_request_args={},
+    ):
+        """Wrapper for _send_request which can optionally retry the request
+        upon receiving a combination of a 400 HTTP response code and a
+        'M_UNRECOGNIZED' errcode. This is a workaround for Synapse <= v0.99.3
+        due to #3622.
+
+        Args:
+            request (MatrixFederationRequest): details of request to be sent
+            try_trailing_slash_on_400 (bool): Whether on receiving a 400
+                'M_UNRECOGNIZED' from the server to retry the request with a
+                trailing slash appended to the request path.
+            404_backoff (bool): Whether to backoff on 404 when making a
+                request with a trailing slash (only affects request if
+                try_trailing_slash_on_400 is True).
+            send_request_args (Dict): A dictionary of arguments to pass to
+                `_send_request()`.
+
+        Raises:
+            HttpResponseException: If we get an HTTP response code >= 300
+                (except 429).
+
+        Returns:
+            Deferred[Dict]: Parsed JSON response body.
+        """
+        try:
+            response = yield self._send_request(**send_request_args)
+
+            # Check if it's necessary to retry with a trailing slash
+            body = yield _handle_json_response(
```

Comment:
I don't think this (or the other instance need to be in here). The exception, if it happens, will get thrown by `_send_request`. Suggest having `_send_request_with_optional_trailing_slash` return the `response`, and leaving the calls to `_handle_json_response` where they were.

Reply:
It will be necessary if we're checking the body for `M_UNRECOGNIZED`, which we need to do again anyways.